capeinfra/pipeline/batch.py

```diff
@@ -0,0 +1,177 @@
+"""Abstractions for batch pipelines."""
+
+from typing import NotRequired, Sequence, TypedDict
+
+import pulumi_aws as aws
+from pulumi import Input, ResourceOptions
+
+from ..iam import get_inline_role, get_instance_profile
+from ..pulumi import DescribedComponentResource
+
+
+# NOTE: Is there a better way to define and maintain validated data structures
+# in python that reesolve to a dictionary and have required/optional fields?
+class BatchComputeResources(TypedDict):
+    """A class for defining resources for instances in an AWS Batch compute environment
+
+    Attributes:
+        instance_types: A list of AWS EC2 instance types to request
+        max_vcpus: The maximum number of vCPUs in an environment
+        desired_vcpus: The desired number of vCPUs in an environment (Optional)
+        min_vcpus: The minimum number of vCPUs in an environment (Optional)
+    """
+
+    instance_types: Sequence[str]
+    max_vcpus: int
+    desired_vcpus: NotRequired[int]
+    min_vcpus: NotRequired[int]
+
+
+def new_batch_compute_resources(
+    instance_types: Sequence[str] = ["c4.large"],
+    max_vcpus: int = 16,
+    desired_vcpus: int | None = None,
+    min_vcpus: int | None = None,
+):
```

Comment:
Dunno how strict we're wanting to be, but there's no docstring for this function.